instance/models/utils.py

```diff
@@ -44,3 +55,263 @@ def save(self, *args, **kwargs):
+                    def wrapped_method(*args, **kwargs):
+                        """ Wrapper around the method which checks the state requirements first """
+                        require_valid_state(resource)
+                        return method(resource, *args, **kwargs)
+                    wrapped_method.is_available = lambda: isinstance(descriptor.__get__(resource), accepted_states)
+                    return wrapped_method
+            return MagicWrapper()
+        return wrap
+
+    def transition(self, to_state, from_states=None):
+        """
+        Returns a "transition" method, which can be called to change the current state from
+        state(s) 'from' to 'to'.
+
+        from_states can be a state class, a tuple of classes, a parent class or mixin of various
+        state classes, a tuple of mixins, etc.. if None, any valid state will be accepted.
+        """
+        assert to_state in self.state_classes
+
+        def do_transition(resource):
+            """ The method that performs a specific transition """
+            current_state = self.__get__(resource)
+            if from_states and not isinstance(current_state, from_states):
+                raise WrongStateException("This transition cannot be used to move from {} to {}".format(
+                    current_state.name, to_state.name  # pylint: disable=no-member
+                ))
+            self._set_state(resource, to_state)
+        do_transition.from_states = from_states  # Convenient way for other code to inspect this transition
+        do_transition.to_state = to_state  # Convenient way for other code to inspect this transition
+        return do_transition
+
+    @property
+    def model_field_choices(self):
+        """
+        Get a tuple of (state_id, name) pairs.
+
+        Suitable for passing to a django CharField choices parameter.
+        """
+        return tuple((state.state_id, state.__class__.__name__) for state in self.state_classes)
+
+    # Internal helper methods:
+
+    def _get_initial_state(self, resource):
+        """
+        Get the initial ResourceState subclass that should be used for this resource.
+        """
+        assert resource not in self.cache, "_get_initial_state is only for determining the initial state."
+        # If the current state was saved into the database in a django field, use that:
+        if self.model_field_name:
+            state_id = getattr(resource, self.model_field_name)
+            if state_id:
+                state_class = self._get_state_class_from_id(state_id)
+                if state_class:
+                    return state_class
+        # Otherwise, just use the default:
+        return self.default_state_class
+
+    def _get_state_class_from_id(self, state_id):
+        """
+        Given a state_id, get the ResourceState subclass, or None
+        """
+        for state_class in self.state_classes:
```

Comment:
My thought was that if you have so many states that this matters, you should refactor your code. But on second thought, it's easy to change to a `frozenset` and that has nicer semantics anyways, so that's what I did.